sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala

```diff
@@ -1408,6 +1409,23 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
       checkDataset(ds, SpecialCharClass("1", "2"))
     }
   }
+
+  test("SPARK-22472: add null check for top-level primitive values") {
+    // If the primitive values are from Option, we need to do runtime null check.
+    val ds = Seq(Some(1), None).toDS().as[Int]
+    intercept[NullPointerException](ds.collect())
+    val e = intercept[SparkException](ds.map(_ * 2).collect())
+    assert(e.getCause.isInstanceOf[NullPointerException])
+
+    withTempPath { path =>
+      Seq(new Integer(1), null).toDF("i").write.parquet(path.getCanonicalPath)
```

Comment:
Is this PR orthogonal to data source format?
Could you test more data source like `JSON`, here?

Reply:
It doesn't matter, I just need a dataframe, I can even use `Seq(Some(1), None).toDF`, but using parquet is more convincing.

Reply:
Thanks!